### ISSUE TITLE
Use one single model module

### ISSUE THREAD

**Issue by rabernat** [m2lines/L96_demo#22]
The original `L96_model.py` file has been copied between every different section. This is a potential source of errors and inconsistencies.  Here is a list of all of the modules in the repo

```
./01Intro/L96_model.py
./02type-of-parametrization/L96_model.py
./03Data-Assimilation/DA_methods.py
./03Data-Assimilation/L96_model.py
./04Subgrid-parametrization-pytorch/L96_model_XYtend.py
./05Offline-DA-increments/DA_methods.py
./05Offline-DA-increments/L96_model.py
./06Different-ML-models/L96_model.py
./06Different-ML-models/L96_model_XYtend.py
./07-Interpretability-of-ML/L96_model_XYtend.py
./07-Interpretability-of-ML/LRP_gradient.py
./08-Implementation/L96_model_XYtend.py
```

We should try to consolidate around a single file.

Steps:
- [ ] Understand the differences between the different L96_model modules in the different sections
- [ ] Create on single "best" version
- [ ] Move everything into a single directory (notebooks and modules)
- [ ] Refactor all the notebooks to use the same model file

Furthermore, we should try to eliminate the use of modules for anything other than L96 itself. This means `L96_model_XYtend.py`, `LRP_gradient.py`, `DA_methods.py`: all of this should be in notebooks, not modules.

**Comment by adcroft** [comment]
Some notes for upcoming PR...

There are three versions of the L96 python module:
```
$ find . -name "L96*.py" | xargs md5sum | sort                                                                                                       
01fb2a1bb99e26d2a8785d7f30b8e368  ./04Subgrid-parametrization-pytorch/L96_model_XYtend.py
01fb2a1bb99e26d2a8785d7f30b8e368  ./06Different-ML-models/L96_model_XYtend.py
01fb2a1bb99e26d2a8785d7f30b8e368  ./07-Interpretability-of-ML/L96_model_XYtend.py
b527fe57fdd668b473c795d18ecf6519  ./01Intro/L96_model.py
b527fe57fdd668b473c795d18ecf6519  ./02type-of-parametrization/L96_model.py
b527fe57fdd668b473c795d18ecf6519  ./03Data-Assimilation/L96_model.py
b527fe57fdd668b473c795d18ecf6519  ./05Offline-DA-increments/L96_model.py
b527fe57fdd668b473c795d18ecf6519  ./06Different-ML-models/L96_model.py
c4d6e77fd3185920d182f7e113cc34eb  ./08-Implementation/L96_model_XYtend.py
```
The smallest difference is between `08-Implementation/L96_model_XYtend.py` and `06Different-ML-models/L96_model.py`:
```
$ diff 08-Implementation/L96_model_XYtend.py 06Different-ML-models/L96_model_XYtend.py                                                               11c11
< def L96_eq1_xdot(X, F,advect=True):
---
> def L96_eq1_xdot(X, F):
26,29c26                                                                                                                
<     if advect:
<         Xdot = np.roll(X,1) * ( np.roll(X,-1) - np.roll(X,2) ) - X + F
<     else:
<         Xdot = - X + F
>     Xdot = np.roll(X,1) * ( np.roll(X,-1) - np.roll(X,2) ) - X + F
```
Replacing all `L96_model_XYtrend.py` with that from `08-Implementation/` seems simple enough.

The larger differences are between `L96_model.py` and `L96_model_XYtrend.py`:
```
diff L96_model.py L96_model_XYtend.py
11c11
< def L96_eq1_xdot(X, F):
---
> def L96_eq1_xdot(X, F, advect=True):
26c26,29
<     Xdot = np.roll(X,1) * ( np.roll(X,-1) - np.roll(X,2) ) - X + F
---
>     if advect:
>         Xdot = np.roll(X,1) * ( np.roll(X,-1) - np.roll(X,2) ) - X + F
>     else:
>         Xdot = - X + F
67c70
<     return Xdot, Ydot
---
>     return Xdot, Ydot, - hcb * Ysummed
186c189
<     time, xhist, yhist = t0+np.zeros((nt+1)), np.zeros((nt+1,len(X0))), np.zeros((nt+1,len(Y0)))
---
>     time, xhist, yhist, xytend_hist = t0+np.zeros((nt+1)), np.zeros((nt+1,len(X0))), np.zeros((nt+1,len(Y0))), np.zeros((nt+1,len(X0)))
189a193
>     xytend_hist[0,:] = 0
199,202c203,206
<             Xdot1,Ydot1 = L96_2t_xdot_ydot(X, Y, F, h, b, c)
<             Xdot2,Ydot2 = L96_2t_xdot_ydot(X+0.5*dt*Xdot1, Y+0.5*dt*Ydot1, F, h, b, c)
<             Xdot3,Ydot3 = L96_2t_xdot_ydot(X+0.5*dt*Xdot2, Y+0.5*dt*Ydot2, F, h, b, c)
<             Xdot4,Ydot4 = L96_2t_xdot_ydot(X+dt*Xdot3, Y+dt*Ydot3, F, h, b, c)
---
>             Xdot1,Ydot1, XYtend = L96_2t_xdot_ydot(X, Y, F, h, b, c)
>             Xdot2,Ydot2, _ = L96_2t_xdot_ydot(X+0.5*dt*Xdot1, Y+0.5*dt*Ydot1, F, h, b, c)
>             Xdot3,Ydot3, _ = L96_2t_xdot_ydot(X+0.5*dt*Xdot2, Y+0.5*dt*Ydot2, F, h, b, c)
>             Xdot4,Ydot4, _ = L96_2t_xdot_ydot(X+dt*Xdot3, Y+dt*Ydot3, F, h, b, c)
206,207c210,211
<         xhist[n+1], yhist[n+1], time[n+1] = X, Y, t0+si*(n+1)
<     return xhist, yhist, time
---
>         xhist[n+1], yhist[n+1], time[n+1],xytend_hist[n+1] = X, Y, t0+si*(n+1), XYtend
>     return xhist, yhist, time, xytend_hist
221a226
>     XYtend = "tendency of X due to small scale vars"
234a240
>         self.XYtend = 0
249c255
<     def set_param(self, dt=None, F=None, h=None, b=None, c=None, t=None):
---
>     def set_param(self, dt=None, F=None, h=None, b=None, c=None, t=0):
276c282
<         X,Y,t = integrate_L96_2t(self.X, self.Y, si, nt, self.F, self.h, self.b, self.c, t0=self.t, dt=self.dt)
---
>         X,Y,t,XYtend = integrate_L96_2t(self.X, self.Y, si, nt, self.F, self.h, self.b, self.c, t0=self.t, dt=self.dt)
279c285
<         return X,Y,t
---
>         return X,Y,t,XYtend
313c319
<     def set_param(self, dt=None, F=None, t=None, method=None):
---
>     def set_param(self, dt=None, F=None, t=0, method=EulerFwd):
```
I suggest we adopt the extended model and then shorten the name back to `L96_mode.py`. The earlier notebooks will have to be updated to ignore the extra data being returned from the extended model.

**Comment by rabernat** [comment]
This was fixed by #32.
